### PR TITLE
NIAD-125: FPN FP69 HA to GP prior notification

### PR DESCRIPTION
**Pull request by bartek-sarul**

### CODE REVIEW

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/FP69ExpiryDate.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+@Builder
+@Data
+public class FP69ExpiryDate extends Segment {
+    private final static String KEY = "DTM";
+    private final static DateTimeFormatter DATE_TIME_FORMATTER = DateTimeFormatter.ofPattern("yyyyMMdd").withZone(TimestampService.UKZone);
+    private final static String QUALIFIER = "962";
+    public final static String KEY_QUALIFIER = KEY + PLUS_SEPARATOR + QUALIFIER;
+    private final static String DATE_FORMAT = "102";
+
+    private @NonNull Instant timestamp;
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/FP69ExpiryDate.java]: Don't use `Instant` when `LocalDate` is enough

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/FP69ReasonCode.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+@Builder
+@Data
+public class FP69ReasonCode extends Segment {
+
+    private final static String KEY = "HEA";
+    private final static String QUALIFIER = "FRN";
+    public final static String KEY_QUALIFIER = KEY + PLUS_SEPARATOR + QUALIFIER;
+    private final static String ZZZ_SUFFIX = ":ZZZ";
+
+    private @NonNull Integer code;
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/FP69ReasonCode.java]: Could be `final`

**Review by grzegorz-robak — CHANGES_REQUESTED** [review]

**Review by grzegorz-robak — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/PersonAddress.java [hunk 3]
```diff
@@ -15,8 +15,10 @@
 @Builder
 @Data
 public class PersonAddress extends Segment {
-    private final static String NAME_AND_ADDRESS = "NAD";
-    private final static String PAT_CODE = "PAT";
+    // NAD+PAT++:::ORPINGTON:'
```
Comment (mattd-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/PersonAddress.java]: stray comment?

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/Segment.java [hunk 4]
```diff
@@ -12,7 +12,7 @@
     protected final static String COLON_SEPARATOR = ":";
     private static final String TERMINATOR = "'";
 
-    protected static <T> List<T> removeEmptyTrailingFields(List<T> list, Predicate<T> predicate) {
+    public static <T> List<T> removeEmptyTrailingFields(List<T> list, Predicate<T> predicate) {
```
Comment (mattd-kainos) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/Segment.java]: why public?

**Review by mattd-kainos — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/Segment.java (hunk 4 above)
Comment (bartek-sarul) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/model/edifact/Segment.java]: ahh, a leftover while I was experimenting

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/service/edifact_to_fhir/FP69PriorNotificationTransactionMapper.java [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            .ifPresent(parameters::addParameter);
+    }
+
+    private void mapAddressLines(PersonAddress personAddress, Address address) {
+        address.setLine(Stream.of(
+            Optional.ofNullable(personAddress.getAddressLine1()),
+            Optional.ofNullable(personAddress.getAddressLine2()),
+            Optional.ofNullable(personAddress.getAddressLine3()),
+            Optional.ofNullable(personAddress.getAddressLine4()),
+            Optional.ofNullable(personAddress.getAddressLine5()))
```
Comment (grzegorz-robak) [inline src/main/java/uk/nhs/digital/nhsconnect/nhais/service/edifact_to_fhir/FP69PriorNotificationTransactionMapper.java:156]: These Optionals are redundant

**Review by grzegorz-robak — DISMISSED** [review]

**Review by mattd-kainos — DISMISSED** [review]

File: src/main/java/uk/nhs/digital/nhsconnect/nhais/service/edifact_to_fhir/FP69PriorNotificationTransactionMapper.java (hunk 5 above)
Comment (grzegorz-robak) [reply src/main/java/uk/nhs/digital/nhsconnect/nhais/service/edifact_to_fhir/FP69PriorNotificationTransactionMapper.java:156]: Ok, I was kind of wrong as `Stream.of()` doesn't support nulls, but... that would work:
````java
address.setLine(Stream.builder()
	.add(personAddress.getAddressLine1())
	.add(personAddress.getAddressLine2())
	.add(personAddress.getAddressLine3())
	.add(personAddress.getAddressLine4())
	.add(personAddress.getAddressLine5())
	.build()
	.map(NullableStringType::new)
    .map(StringType.class::cast)
    .collect(Collectors.toList()));
````

**Review by Alicja-Kainos — DISMISSED** [review]

**Review by grzegorz-robak — APPROVED** [review]

**Pull request merged**